presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonOperators.java

```diff
@@ -13,20 +13,153 @@
 import static com.facebook.presto.metadata.OperatorType.HASH_CODE;
 import static com.facebook.presto.metadata.OperatorType.NOT_EQUAL;
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_CAST_ARGUMENT;
+import static com.facebook.presto.spi.type.StandardTypes.JSON;
+import static com.fasterxml.jackson.core.JsonFactory.Feature.CANONICALIZE_FIELD_NAMES;
+import static java.lang.String.format;
 
 public final class JsonOperators
 {
+    private static final JsonFactory JSON_FACTORY = new JsonFactory().disable(CANONICALIZE_FIELD_NAMES);
+
     private JsonOperators()
     {
     }
 
+    @ScalarOperator(OperatorType.CAST)
+    @SqlType(StandardTypes.VARCHAR)
+    public static Slice castToVarchar(@SqlType(StandardTypes.JSON) Slice slice)
+    {
+        return slice;
+    }
+
+    @ScalarOperator(OperatorType.CAST)
+    @SqlType(StandardTypes.BIGINT)
+    public static long castToBigint(@SqlType(StandardTypes.JSON) Slice json)
+    {
+        try {
+            try (JsonParser parser = JSON_FACTORY.createJsonParser(json.getInput())) {
+                parser.nextToken();
+                return parser.getLongValue();
+            }
+        }
+        catch (Exception e) {
+            throw new PrestoException(INVALID_CAST_ARGUMENT, format("Cannot cast '%s' to %s", json.toStringUtf8(), StandardTypes.BIGINT));
+        }
+    }
+
+    @ScalarOperator(OperatorType.CAST)
+    @SqlType(StandardTypes.DOUBLE)
+    public static double castToDouble(@SqlType(StandardTypes.JSON) Slice json)
+    {
+        try {
+            try (JsonParser parser = JSON_FACTORY.createJsonParser(json.getInput())) {
+                parser.nextToken();
+                return parser.getDoubleValue();
+            }
+        }
+        catch (Exception e) {
+            throw new PrestoException(INVALID_CAST_ARGUMENT, format("Cannot cast '%s' to %s", json.toStringUtf8(), StandardTypes.DOUBLE));
+        }
+    }
+
+    @ScalarOperator(OperatorType.CAST)
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean castToBoolean(@SqlType(StandardTypes.JSON) Slice json)
+    {
+        try {
+            try (JsonParser parser = JSON_FACTORY.createJsonParser(json.getInput())) {
+                parser.nextToken();
+                return parser.getBooleanValue();
+            }
+        }
+        catch (Exception e) {
+            throw new PrestoException(INVALID_CAST_ARGUMENT, format("Cannot cast '%s' to %s", json.toStringUtf8(), StandardTypes.BOOLEAN));
+        }
+    }
+
+    @ScalarOperator(OperatorType.CAST)
+    @SqlType(StandardTypes.JSON)
+    public static Slice castToJson(@SqlType(StandardTypes.VARCHAR) Slice slice) throws IOException
+    {
+        try {
+            byte[] input = slice.getBytes();
+            SliceOutput dynamicSliceOutput = new DynamicSliceOutput(input.length);
+            try (JsonGenerator jsonGenerator = JSON_FACTORY.createJsonGenerator(dynamicSliceOutput)) {
+                jsonGenerator.writeRaw(slice.toStringUtf8());
```

Comment:
I think the problem is the cast is mixing two things together.  We have the concept of we have json encoded stuff in a VARCHAR, and we want to make it type JSON so we can use json functions on it.  Then so the opposite cast, should take structural JSON and turn it into VARCHAR.  That would mean we would, simply return the json text.  I think what you are proposing is that we would expect the json to contain a single String value (in quotes) and we would read just that value.  In the spec you link above it states `A string is a sequence of Unicode code points wrapped with quotation marks (U+0022)`.